app/models/data_export.rb

```diff
@@ -0,0 +1,79 @@
+require "csv"
+
+class DataExport
+  SUPPORTED_TYPES = %w(
+    Donation
+    DonationSite
+    Partner
+    Purchase
+    Distribution
+    DiaperDriveParticipant
+    StorageLocation
+    Adjustment
+    Transfer
+    Item
+    BarcodeItem
+  ).freeze
+
+  def initialize(organization, type)
+    @current_organization = organization
+    @type = type
+  end
+
+  def as_csv
+    return nil unless current_organization.present? && type.present?
+    return nil unless SUPPORTED_TYPES.include? type
+
+    data_to_export = grab_data_to_export
+    headers = type.constantize.csv_export_headers
+    generate_csv(data_to_export, headers)
+  end
+
+  private
+
+  attr_reader :current_organization, :type
+
+  def grab_data_to_export
+    case type
+    when "Donation"
+      current_organization.donations
```

Comment:
I like it how it is, but an alternative would be to introduce a `scope` or method to each exportable model that locally declared it's own includes/order, parameterized by the organization.

    # in Donation model
    scope :for_csv_export, -> (organization) {
      where(organization: organization)
      .includes(:line_items, :storage_location, :donation_site)
      .order(created_at: :desc)
    }

Then something like `type.constantize.for_csv_export(current_organization` instead of the case blocks.

Reply:
I really like this! I'll make the change soon